src/Avalonia.Base/Input/FocusManager.cs

```diff
@@ -58,94 +52,59 @@ static FocusManager()
             NavigationMethod method = NavigationMethod.Unspecified,
             KeyModifiers keyModifiers = KeyModifiers.None)
         {
-            if (control != null)
+            if (KeyboardDevice.Instance is not { } keyboardDevice)
+                return false;
+
+            if (control is not null)
             {
-                var scope = GetFocusScopeAncestors(control)
-                    .FirstOrDefault();
+                if (!CanFocus(control))
+                    return false;
 
-                if (scope != null)
+                if (GetFocusScope(control) is StyledElement scope)
                 {
-                    Scope = scope;
-                    return SetFocusedElement(scope, control, method, keyModifiers);
+                    scope.SetValue(FocusedElementProperty, control);
+                    _focusRoot = GetFocusRoot(scope);
                 }
+
+                keyboardDevice.SetFocusedElement(control, method, keyModifiers);
+                return true;
             }
-            else if (Current != null)
+            else if (_focusRoot?.GetValue(FocusedElementProperty) is { } restore && 
```

Comment:
I find the _focusRoot concept a bit confusing. Wouldn't it be better to just stick to focus scopes like in the old implementation?

At least in theory there could be another focus scope between the cleared scope and the VisualRoot.

Reply:
This is just following [WPF's behavior](https://github.com/AvaloniaUI/Avalonia/issues/13325#issuecomment-1772655137) of focusing the outermost focus scope when a focused control is removed. The reason we need to store the focus root is that at the time the focused control is removed we may not have all the information we need to find that outermost scope.